File: lua/test/route_pattern_entity_test.lua
-- RoutePattern entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("mbta-v3_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("RoutePatternEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:RoutePattern(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = route_pattern_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "route_pattern." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set MBTAV__TEST_ROUTE_PATTERN_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local route_pattern_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.route_pattern")))
    local route_pattern_ref01_data = nil
    if #route_pattern_ref01_data_raw > 0 then
      route_pattern_ref01_data = helpers.to_map(route_pattern_ref01_data_raw[1][2])
    end

    -- LOAD
    local route_pattern_ref01_ent = client:RoutePattern(nil)
    local route_pattern_ref01_match_dt0 = {}
    local route_pattern_ref01_data_dt0_loaded, err = route_pattern_ref01_ent:load(route_pattern_ref01_match_dt0, nil)
    assert.is_nil(err)
    assert.is_not_nil(route_pattern_ref01_data_dt0_loaded)

  end)
end)

function route_pattern_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/route_pattern/RoutePatternTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read route_pattern test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "route_pattern01", "route_pattern02", "route_pattern03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("MBTAV__TEST_ROUTE_PATTERN_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["MBTAV__TEST_ROUTE_PATTERN_ENTID"] = idmap,
    ["MBTAV__TEST_LIVE"] = "FALSE",
    ["MBTAV__TEST_EXPLAIN"] = "FALSE",
    ["MBTAV__APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["MBTAV__TEST_ROUTE_PATTERN_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["MBTAV__TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["MBTAV__APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["MBTAV__TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["MBTAV__TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/route_pattern_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/mbta-v3-sdk/go"
	"github.com/voxgig-sdk/mbta-v3-sdk/go/core"

	vs "github.com/voxgig-sdk/mbta-v3-sdk/go/utility/struct"
)

func TestRoutePatternEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.RoutePattern(nil)
		if ent == nil {
			t.Fatal("expected non-nil RoutePatternEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := route_patternBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "route_pattern." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set MBTA_V3_TEST_ROUTE_PATTERN_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		routePatternRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.route_pattern", setup.data)))
		var routePatternRef01Data map[string]any
		if len(routePatternRef01DataRaw) > 0 {
			routePatternRef01Data = core.ToMapAny(routePatternRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = routePatternRef01Data

		// LOAD
		routePatternRef01Ent := client.RoutePattern(nil)
		routePatternRef01MatchDt0 := map[string]any{}
		routePatternRef01DataDt0Loaded, err := routePatternRef01Ent.Load(routePatternRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if routePatternRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func route_patternBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "route_pattern", "RoutePatternTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read route_pattern test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse route_pattern test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"route_pattern01", "route_pattern02", "route_pattern03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("MBTA_V3_TEST_ROUTE_PATTERN_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"MBTA_V3_TEST_ROUTE_PATTERN_ENTID": idmap,
		"MBTA_V3_TEST_LIVE":      "FALSE",
		"MBTA_V3_TEST_EXPLAIN":   "FALSE",
		"MBTA_V3_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["MBTA_V3_TEST_ROUTE_PATTERN_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["MBTA_V3_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["MBTA_V3_APIKEY"],
			},
			extra,
		})
		client = sdk.NewMbtaV3SDK(core.ToMapAny(mergedOpts))
	}

	live := env["MBTA_V3_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["MBTA_V3_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
